pkg/phlaredb/block_querier.go

```diff
@@ -771,54 +809,90 @@ func MergeProfilesStacktraces(ctx context.Context, stream *connect.BidiStream[in
 		otlog.String("profile_id", request.Type.ID),
 	)
 
-	queriers, err := blockGetter(ctx, model.Time(request.Start), model.Time(request.End))
+	queriers, err := blockGetter(ctx, model.Time(request.Start), model.Time(request.End), request.Hints)
 	if err != nil {
 		return err
 	}
 
-	iters, err := SelectMatchingProfiles(ctx, request, queriers)
-	if err != nil {
-		return err
-	}
-
-	// send batches of profiles to client and filter via bidi stream.
-	selectedProfiles, err := filterProfiles[
-		BidiServerMerge[*ingestv1.MergeProfilesStacktracesResponse, *ingestv1.MergeProfilesStacktracesRequest],
-		*ingestv1.MergeProfilesStacktracesResponse,
-		*ingestv1.MergeProfilesStacktracesRequest](ctx, iters, defaultBatchSize, stream)
-	if err != nil {
-		return err
+	deduplicationNeeded := true
+	if request.Hints != nil && request.Hints.Block != nil {
+		deduplicationNeeded = request.Hints.Block.Deduplication
 	}
 
 	var m sync.Mutex
 	t := new(phlaremodel.Tree)
 	g, ctx := errgroup.WithContext(ctx)
 
-	for i, querier := range queriers {
-		querier := querier
-		i := i
-		if len(selectedProfiles[i]) == 0 {
-			continue
+	// depending on if new need deduplication or not there are two different code paths.
+	if !deduplicationNeeded {
+		// in this path we can just merge the profiles from each block and send the result to the client.
+		for _, querier := range queriers {
+			querier := querier
+			g.Go(util.RecoverPanic(func() error {
+				// signal the end of the profile streaming by sending an empty response.
+				sp.LogFields(otlog.String("msg", "no profile streaming as no deduplication needed"))
+				if err = stream.Send(&ingestv1.MergeProfilesStacktracesResponse{}); err != nil {
+					return err
+				}
+
+				iters, err := querier.SelectMatchingProfiles(ctx, request)
+				if err != nil {
+					return err
+				}
+
+				// TODO: To sort or not?
```

Comment:
You need sorting if we don't change how `SelectMatchingProfiles` orders data, currently it reorders by time and then series.  

@kolesnikovae wanted to change this, and I couldn't remember if this was important or not.

Reply:
This is fixed in the meantime